File: lib/utils/getBoundsFromNodeWithPortPoints.ts
import { NodeWithPortPoints } from "lib/types/high-density-types"
import { getNodeBounds } from "./capacityMeshNodeGeometry"

export function getBoundsFromNodeWithPortPoints(
  nodeWithPortPoints: NodeWithPortPoints,
): { minX: number; maxX: number; minY: number; maxY: number } {
  const bounds = { ...getNodeBounds(nodeWithPortPoints as any) }

  // Sometimes port points may be outside the node- this happens when there's
  // a "leap" to the final target or at the end or beginning of a trace when
  // we're wrapping up
  for (const pt of nodeWithPortPoints.portPoints) {
    if (pt.x < bounds.minX) {
      bounds.minX = pt.x
    }
    if (pt.x > bounds.maxX) {
      bounds.maxX = pt.x
    }
    if (pt.y < bounds.minY) {
      bounds.minY = pt.y
    }
    if (pt.y > bounds.maxY) {
      bounds.maxY = pt.y
    }
  }

  return bounds
}

File: rust/capacity-autorouter/src/solvers/uniform_port_distribution_solver/get_bounds_from_node_with_port_points.rs
use crate::solvers::uniform_port_distribution_solver::types::{Bounds, NodeWithPortPoints};

pub fn get_bounds_from_node_with_port_points(node: &NodeWithPortPoints) -> Bounds {
    let mut bounds = Bounds {
        min_x: node.center.x - node.width / 2.0,
        max_x: node.center.x + node.width / 2.0,
        min_y: node.center.y - node.height / 2.0,
        max_y: node.center.y + node.height / 2.0,
    };
    for point in &node.port_points {
        if point.x < bounds.min_x { bounds.min_x = point.x; }
        if point.x > bounds.max_x { bounds.max_x = point.x; }
        if point.y < bounds.min_y { bounds.min_y = point.y; }
        if point.y > bounds.max_y { bounds.max_y = point.y; }
    }
    bounds
}
